### PR TITLE
[WIP] Used  per-parameter FSDP

### PR DESCRIPTION
**Pull request by awgu**
This PR shows how we would modify torchtrain to use per-parameter FSDP (sometimes called FSDP2 or ppFSDP). Detailed context can be found in https://github.com/pytorch/pytorch/issues/114299.

**Progress Tracking**
Gaps for 1D FSDP:
- Support gradient norm clipping
    - The plan is to rely on `DTensor`'s op dispatch to implement `vector_norm` and `stack`. This approach will allow us to use `torch.nn.utils.clip_grad_norm_()` out of the box as long as all `DTensor` parameters share the same sharding.
    - [x] Gradient norm clipping w/o `foreach=True`: https://github.com/pytorch/pytorch/pull/120238
    - [x] Gradient norm clipping w/ `foreach=True`
- [ ] Integrate with distributed checkpointing
- [ ] Meta-device initialization
    - The plan is to allow FSDP to shard on meta device without materializing parameters. After applying FSDP, the user can call `model.to_empty(device="cuda")` and initialize parameters/buffers as needed (e.g. via `reset_parameters()` on each module). The initialization will use `DTensor`'s randomness, which ensures correct randomness with respect to the global shape.
    - There are some minor gaps in core that will be fixed soon.

Goal is to land PRs to address these gaps by 3/8. The overall execution tracker can be found in https://github.com/pytorch/pytorch/issues/120003.

Gaps for 2D FSDP + TP/SP:
- [ ] Support `torch.compile`
    - Current error: P1192664996
    - We may need to add FSDP2 to some skipfiles similar to FSDP so that we graph break cleanly on FSDP's pre-forward hook.

**Experiments**
LLama2-7B
8 H100 GPUs, local batch size 8, sequence length 2048, activation checkpointing and `torch.compile` each transformer block
- Baseline: Existing flat-parameter (`FULL_SHARD`):
  - 2124 ms per iteration
  - 23.24 GB peak active
- Per-parameter (`reshard_after_forward=True`)
  - 2143 ms per iteration
  - 22.11 GB peak active
- Per-parameter (`reshard_after_forward=False` for last transformer block)
  - 2127 ms per iteration
  - 22.49 GB peak active (if peak is at beginning of backward, then this increase makes sense)

(Ideally, we can have some per-GPU MFU or tokens-per-second metric rather than time per iteration, which is less interpretable.)

### CODE REVIEW

File: torchtrain/parallelisms/parallelize_llama.py [hunk 1]
```diff
@@ -157,32 +151,15 @@ def parallelize_llama(model, world_mesh, parallel_dims, args):
...
-
-            # wrap the rest layers with FSDP
-            model = wrap(model.cuda())
-
+        mp_policy = MixedPrecisionPolicy(
+            param_dtype=torch.bfloat16, reduce_dtype=torch.float32
+        )
+        fsdp_config = {"mesh": dp_mesh, "mp_policy": mp_policy}
+        for layer_id, transformer_block in enumerate(model.layers):
+            transformer_block = checkpoint_wrapper(transformer_block, args)
```
Comment (awgu) [inline torchtrain/parallelisms/parallelize_llama.py]: If we could get `_composable.checkpoint` to work with compile / selective AC, then we could replace with
```
checkpoint(transformer_block)
```
and not require the assignment `model.layers[layer_id] = transformer_block`.

File: run_llama_train.sh [hunk 2]
```diff
@@ -9,7 +9,7 @@ TRAINER_DIR=${1:-/home/$USER/local/torchtrain}
 # LOG_RANK=0,1 NGPU=4 SP=2 ./run_llama_train.sh
 
 MODEL=${MODEL:-"llama"}
-MODEL_CONF=${MODEL_CONF:-"debugmodel"}
+MODEL_CONF=${MODEL_CONF:-"7B"}
```
Comment (lessw2020) [inline run_llama_train.sh]: was this change to 7B as new default intentional?  
I think we want to stick with 'debugmodel' as default so people don't launch a larger job up front for a first run.

File: train.py [hunk 3]
```diff
@@ -187,7 +178,7 @@ def main(args):
 
             # clip gradients (after unscaling gradients of the optimizer's params)
             scaler.unscale_(optimizer)
-            model.clip_grad_norm_(args.max_norm)
+            # model.clip_grad_norm_(args.max_norm)
```
Comment (lessw2020) [inline train.py]: nit - should probably single warn earlier (not during training loop) that clip_grad_norm is not supported atm, or at least stick a TODO here so we don't forget about this not being enabled (assuming it will be working in the future).

File: run_llama_train.sh (hunk 2 above)
Comment (awgu) [reply run_llama_train.sh]: Good call! This was not intentional.

This PR is not ready to land yet. I will let you guys know once everything is ready. I mainly wanted to open it just to track progress.

File: torchtrain/parallelisms/parallelize_llama.py [hunk 4]
```diff
@@ -152,6 +147,7 @@ def parallelize_llama(model, world_mesh, parallel_dims, job_config: JobConfig):
                 ),
             },
         )
+        distribute_rmsnorm(model.norm, tp_mesh)
```
Comment (awgu) [inline torchtrain/parallelisms/parallelize_llama.py:150]: cc @wanchaol @tianyu-l:
FSDP2 currently checks that if it uses a mesh that has a parent mesh, then the parameter to shard should be a `DTensor`. However, without this `distribute_rmsnorm(model.norm, tp_mesh)`, the `Transformer.norm.weight` is a `torch.Tensor`.

Would the sharded state dict for existing FSDP be incorrect without making it a `DTensor`? Should we assert that all parameters in the model are `DTensor` after parallelizing with TP/SP, or is that not a requirement? (If not a requirement, then we may need to relax the check on FSDP2 side.)
